source/joker_effects.c

```diff
@@ -5,180 +5,206 @@
     u8 suits[NUM_SUITS];
     u8 ranks[NUM_RANKS];
     get_played_distribution(ranks, suits);
 
     if (hand_contains_n_of_a_kind(ranks) >= 2)
         effect.mult = 8;
+
     return effect;
 }
 
-static JokerEffect zany_joker_effect(Joker *joker, Card *scored_card) {
+static JokerEffect zany_joker_effect(Joker *joker, Card *scored_card, int scored_when) {
     JokerEffect effect = {0};
-    if (scored_card != NULL)
-        return effect; // if card != null, we are not at the end-phase of scoring yet
 
+    SCORE_INDEPENDANT_ONLY(scored_card, scored_when, effect)
+
+    // This is really inefficient but the only way at the moment to check for whole-hand conditions
     u8 suits[NUM_SUITS];
     u8 ranks[NUM_RANKS];
     get_played_distribution(ranks, suits);
 
     if (hand_contains_n_of_a_kind(ranks) >= 3)
         effect.mult = 12;
+
     return effect;
 }
 
-static JokerEffect mad_joker_effect(Joker *joker, Card *scored_card) {
+static JokerEffect mad_joker_effect(Joker *joker, Card *scored_card, int scored_when) {
     JokerEffect effect = {0};
-    if (scored_card != NULL)
-        return effect; // if card != null, we are not at the end-phase of scoring yet
+
+    SCORE_INDEPENDANT_ONLY(scored_card, scored_when, effect)
 
     u8 suits[NUM_SUITS];
     u8 ranks[NUM_RANKS];
     get_played_distribution(ranks, suits);
 
     if (hand_contains_two_pair(ranks))
         effect.mult = 10;
+
     return effect;
 }
 
-static JokerEffect crazy_joker_effect(Joker *joker, Card *scored_card) {
+static JokerEffect crazy_joker_effect(Joker *joker, Card *scored_card, int scored_when) {
     JokerEffect effect = {0};
-    if (scored_card != NULL)
-        return effect; // if card != null, we are not at the end-phase of scoring yet
+
+    SCORE_INDEPENDANT_ONLY(scored_card, scored_when, effect)
 
     u8 suits[NUM_SUITS];
     u8 ranks[NUM_RANKS];
     get_played_distribution(ranks, suits);
 
     if (hand_contains_straight(ranks))
         effect.mult = 12;
+
     return effect;
 }
 
-static JokerEffect droll_joker_effect(Joker *joker, Card *scored_card) {
+static JokerEffect droll_joker_effect(Joker *joker, Card *scored_card, int scored_when) {
     JokerEffect effect = {0};
-    if (scored_card != NULL)
-        return effect; // if card != null, we are not at the end-phase of scoring yet
+
+    SCORE_INDEPENDANT_ONLY(scored_card, scored_when, effect)
 
     u8 suits[NUM_SUITS];
     u8 ranks[NUM_RANKS];
     get_played_distribution(ranks, suits);
 
     if (hand_contains_flush(suits))
         effect.mult = 10;
+
     return effect;
 }
 
-static JokerEffect sly_joker_effect(Joker *joker, Card *scored_card) {
+static JokerEffect sly_joker_effect(Joker *joker, Card *scored_card, int scored_when) {
     JokerEffect effect = {0};
-    if (scored_card != NULL)
-        return effect; // if card != null, we are not at the end-phase of scoring yet
+
+    SCORE_INDEPENDANT_ONLY(scored_card, scored_when, effect)
```

Comment:
```suggestion
    SCORE_ON_EVENT_ONLY(JOKER_CALLBACK_INDEPENDANT, scored_when, effect)
```

Batch - Consolidate joker event restriction macros